src/repart/repart.c

```diff
@@ -5824,76 +5824,80 @@ static int add_subvolume_path(const char *path, Set **subvolumes) {
+        _cleanup_hashmap_free_ Hashmap *hashmap = NULL;
+        _cleanup_free_ char *path = NULL;
         Subvolume *subvolume;
         int r;
 
         assert(p);
         assert(ret);
 
-        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes)
-                if (strv_extend(&subvolumes, subvolume->path) < 0)
+        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes) {
+                path = strdup(subvolume->path);
+                if (!path)
                         return log_oom();
 
-        if (p->suppressing) {
-                char **suppressing;
-
-                r = make_subvolumes_strv(p->suppressing, &suppressing);
-                if (r < 0)
-                        return r;
-
-                r = strv_extend_strv_consume(&subvolumes, suppressing, /* filter_duplicates= */ true);
+                r = hashmap_ensure_put(&hashmap, &string_hash_ops_free, path, UINT_TO_PTR(subvolume->flags));
                 if (r < 0)
                         return log_oom();
+                if (r == 0)
+                        log_debug("Found duplicate subvolume path %s for the same partition", path);
+                TAKE_PTR(path);
         }
 
-        *ret = TAKE_PTR(subvolumes);
+        if (p->suppressing)
+                if (make_subvolumes_hashmap(p->suppressing, &hashmap) < 0)
+                        return r;
+
+        *ret = TAKE_PTR(hashmap);
         return 0;
 }
 
-static int make_subvolumes_set(
+static int make_subvolumes_by_source_inode_hashmap(
                 const Partition *p,
                 const char *source,
                 const char *target,
-                Set **ret) {
+                Hashmap **ret) {
 
-        _cleanup_strv_free_ char **paths = NULL;
-        _cleanup_set_free_ Set *subvolumes = NULL;
+        _cleanup_hashmap_free_ Hashmap *hashmap = NULL;
+        Subvolume *subvolume;
         int r;
 
         assert(p);
         assert(target);
         assert(ret);
 
-        r = make_subvolumes_strv(p, &paths);
-        if (r < 0)
-                return r;
-
-        STRV_FOREACH(subvolume, paths) {
+        ORDERED_HASHMAP_FOREACH(subvolume, p->subvolumes) {
                 _cleanup_free_ char *path = NULL;
 
-                const char *s = path_startswith(*subvolume, target);
+                const char *s = path_startswith(subvolume->path, target);
                 if (!s)
                         continue;
 
                 path = path_join(source, s);
                 if (!path)
                         return log_oom();
 
-                r = add_subvolume_path(path, &subvolumes);
+                r = add_subvolume_path(path, subvolume->flags, &hashmap);
                 if (r < 0)
                         return r;
         }
 
-        *ret = TAKE_PTR(subvolumes);
+        if (p->suppressing)
+                if (make_subvolumes_by_source_inode_hashmap(p->suppressing, source, target, &hashmap) < 0)
+                        return r;
```

Comment:
Same problem here, we can't just pass the existing hashmap as is here. But this function not handling `p->suppressing` is a pre-existing problem, so you don't have to fix this in your PR.